frontend/src/InputForm.js

```diff
@@ -0,0 +1,546 @@
+    this.updateCourseIsRequired = this.updateCourseIsRequired.bind(this);
+    this.updateCourseRank = this.updateCourseRank.bind(this);
+    this.updateRankSelectOptions = this.updateRankSelectOptions.bind(this);
+    
+    // sections
+    this.updateSectionProfessor = this.updateSectionProfessor.bind(this);
+    this.updateSectionStartTime = this.updateSectionStartTime.bind(this);
+    this.updateSectionEndTime = this.updateSectionEndTime.bind(this);
+    this.updateSectionDays = this.updateSectionDays.bind(this);
+
+    // criterion
+    this.updateTimeStartPreference = this.updateTimeStartPreference.bind(this);
+    this.updateTimeEndPreference = this.updateTimeEndPreference.bind(this);
+    this.updateSubjectPreference = this.updateSubjectPreference.bind(this);
+    
+    // basic info
+    this.updateMinCredits = this.updateMinCredits.bind(this);
+    this.updateMaxCredits = this.updateMaxCredits.bind(this);
+    this.updateTermStartDate = this.updateTermStartDate.bind(this);
+    this.updateTermEndDate = this.updateTermEndDate.bind(this);
+
+    this.convertCourseSections = this.convertCourseSections.bind(this);
+    this.submit = this.submit.bind(this);
+  }
+
+  // used to keep track of which ranks have been selected for ordering the courses,
+  // so that no two courses have the same rank 
+  selectedRanks = [];
+
+  updateCourseName(id, courseName) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, name: courseName })
+          : course)
+    });
+  }
+
+  updateCourseID(id, courseID) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, courseID: courseID })
+          : course)
+    });
+  }
+
+  updateCourseSubject(id, courseSubject) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, subject: courseSubject })
+          : course)
+    });
+  }
+
+  updateCourseCredits(id, courseCredits) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, credits: courseCredits })
+          : course)
+    });
+  }
+
+  updateCourseIsRequired(id) {
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.map((course, index) =>
+        index === id
+          ? ({ ...course, isRequired: !this.state.courses[index].isRequired })
+          : course)
+    });
+  }
+
+  updateCourseRank(id, selected) {
+    var previousSelected = this.state.courses[id].rank;
+    if (previousSelected !== "") {
+      var index = this.selectedRanks.indexOf(previousSelected);
+      this.selectedRanks.splice(index, 1);
+    }
+    if (this.selectedRanks.indexOf(selected) === -1) {
+      this.selectedRanks.push(selected);
```

Comment:
Good to know, I'll make that improvement!

Reply:
(Change doesn't apply with new ranking strategy). Also, I am going to make the type="number" change in the next branch, labSections, so that it's applied everywhere